app/services/kredits.js

```diff
@@ -283,29 +300,40 @@ export default Service.extend({
       });
   },
 
-  addProposal(proposal) {
-    const {
-      recipientAddress,
-      amount,
-      url
-    } = proposal.getProperties('recipientAddress', 'amount', 'url');
+  addProposal(attributes) {
+    debug('[kredits] add proposal', attributes);
+
+    let json = ContributionSerializer.serialize(attributes);
+    // TODO: validate against schema
 
     return this.get('ipfs')
-      .storeFile(proposal.serializeContribution())
-      .then(ipfsHash => {
+      .storeFile(json)
+      // Set ipfsHash
+      .then((ipfsHash) => {
+        delete attributes.contributorIpfsHash;
+        attributes.ipfsHash = ipfsHash;
+        return attributes;
+      })
+      .then((attributes) => {
         return this.get('kreditsContract')
           .then((contract) => {
-            return contract.addProposal(
-              recipientAddress,
+            let { recipientId, amount, ipfsHash } = attributes;
+            let { digest, hashFunction, hashSize } = toBytes32(ipfsHash);
+
+            let proposal = [
+              recipientId,
               amount,
-              url,
-              ipfsHash
-            );
-          })
-          .then((data) => {
-            debug('[kredits] add proposal response', data);
-            return data;
+              digest,
+              hashFunction,
+              hashSize,
```

Comment:
I think the hash splitting and combining should not be done in here, but needs to in our kredits npm module, so we can use it from other programs as well.

Reply:
What we try right now is that we cleanup the kredits service and and extract all the common logic from the service in the npm module. That will be our next step.

Reply:
Cool.